### PR TITLE
Spatial grid for neighbor detection

### LINKED ISSUE #174: Performance improvement with Spatial Hashmap
A Redditor suggested me to use a spatial hash map or a k-d tree to improve general performance. One of these methods should replace at least the main loop, and probably some others if I'm right.

The comment is this:
https://www.reddit.com/r/javascript/comments/fmdn6e/tsparticles_what_particlesjs_could_be_if_it_wasnt/flfrqk2?utm_source=share&utm_medium=web2x

There's some articles and code to check.

### PR DESCRIPTION
**Pull request by TylerBurnett**
SpatialGrid.ts 
- Added spatial grid class with respective functions

Particles.ts
- Added SpatialGrid class, initializer in the constructor. 
- Note Grid fills and resets on frame redraw.

Particle.ts 
- Replaced collision detection loop with spatial grid call 
- Refactored redundant radius * 2 calls to const called 'diameter'.

### CODE REVIEW

**Comment by TylerBurnett** [comment]
Looking good, I'll move this forward.

File: src/Classes/Particles.ts [hunk 1]
```diff
@@ -25,6 +27,7 @@ export class Particles {
         this.container = container;
         this.array = [];
         this.interactionsEnabled = false;
+        this.spatialGrid = new SpatialGrid(this.container.canvas.dimension, 20);
```
Comment (matteobruni) [inline src/Classes/Particles.ts:30]: Why 20? Use Constants class for it, so it's readable

File: src/Classes/Particle/Linker.ts [hunk 2]
```diff
@@ -62,4 +62,4 @@ export class Linker {
             }
         }
     }
-}
+}
```
Comment (matteobruni) [inline src/Classes/Particle/Linker.ts:65]: keep the newline at the end of file, Codacy will drop quality for it

**Review by matteobruni — CHANGES_REQUESTED** [review]

**Comment by matteobruni** [comment]
GH is reporting me conflicts:

Conflicting files
src/Classes/Particle.ts
src/Classes/Particles.ts

File: src/Classes/Particle/Linker.ts (hunk 2 above)
Comment (TylerBurnett) [reply src/Classes/Particle/Linker.ts:65]: No worries, my linter must've changed it

File: src/Classes/Particles.ts (hunk 1 above)
Comment (TylerBurnett) [reply src/Classes/Particles.ts]: Moved back into the spatial grid class as preset param.

File: src/Classes/Utils/SpatialGrid.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+*/
+
+export class SpatialGrid {
+    private cellSize: number;
+    private widthSegment: number;
+    private heightSegment: number;
+    private grid: object[][][] = [];
+
+    // Cut the grid up into a 2d array with a 3rd dimesion holding the data.
+    constructor(canvas: IDimension, cellSize = 20) {
```
Comment (matteobruni) [inline src/Classes/Utils/SpatialGrid.ts:24]: Codacy Issue found: [expected parameter: 'cellSize' to have a typedef](https://app.codacy.com/manual/ar3s/tsparticles/pullRequest?prid=5343253)

**Comment by matteobruni** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ar3s/tsparticles/pullRequest?prid=5343253&bid=17493703)

**Review by matteobruni — APPROVED** [review]

**Comment by matteobruni** [comment]
Everything seems ok

**Pull request merged**